lightning-persister/src/lib.rs

```diff
@@ -0,0 +1,263 @@
+extern crate lightning;
+extern crate bitcoin;
+extern crate libc;
+
+use bitcoin::hashes::hex::ToHex;
+use lightning::chain::channelmonitor::{ChannelMonitor, ChannelMonitorUpdate, ChannelMonitorUpdateErr};
+use lightning::chain::channelmonitor;
+use lightning::chain::keysinterface::ChannelKeys;
+use lightning::chain::transaction::OutPoint;
+use lightning::util::ser::{Writeable, Readable};
+use std::fs;
+use std::io::Error;
+use std::marker::PhantomData;
+use std::path::Path;
+
+#[cfg(test)]
+use bitcoin::{BlockHash, Txid};
+use bitcoin::hashes::hex::FromHex;
+use std::collections::HashMap;
+use std::io::Cursor;
+
+#[cfg(not(target_os = "windows"))]
+use std::os::unix::io::AsRawFd;
+
+/// FilesystemPersister persists channel data on disk, where each channel's
+/// data is stored in a file named after its funding outpoint.
+///
+/// Warning: this module does the best it can with calls to persist data, but it
+/// can only guarantee that the data is passed to the drive. It is up to the
+/// drive manufacturers to do the actual persistence properly, which they often
+/// don't (especially on consumer-grade hardware). Therefore, it is up to the
+/// user to validate their entire storage stack, to ensure the writes are
+/// persistent.
+/// Corollary: especially when dealing with larger amounts of money, it is best
+/// practice to have multiple channel data backups and not rely only on one
+/// FilesystemPersister.
+pub struct FilesystemPersister {
+	path_to_channel_data: String,
+}
+
+trait DiskWriteable {
+	fn write(&self, writer: &mut fs::File) -> Result<(), Error>;
+}
+
+impl<ChanSigner: ChannelKeys + Writeable> DiskWriteable for ChannelMonitor<ChanSigner> {
+	fn write(&self, writer: &mut fs::File) -> Result<(), Error> {
+		self.write_for_disk(writer)
+	}
+}
+
+impl FilesystemPersister {
+	/// Initialize a new FilesystemPersister and set the path to the individual channels'
+	/// files.
+	pub fn new(path_to_channel_data: String) -> Self {
+		return Self {
+			path_to_channel_data,
+		}
+	}
+
+	fn get_full_filepath(&self, funding_txo: OutPoint) -> String {
+		let path = Path::new(&self.path_to_channel_data);
+		let mut path_buf = path.to_path_buf();
+		path_buf.push(format!("{}_{}", funding_txo.txid.to_hex(), funding_txo.index));
+		path_buf.to_str().unwrap().to_string()
+	}
+
+	// Utility to write a file to disk.
+	// Note: we may eventually want to change this model to a queue-based system, such that
```

Comment:
I don't think it would help - we *have* to fsync before we return or we violate the requirements of persistence before moving a channel forwards. IIRC various online sources have claimed that you can fsync in a background thread and it improves things, but I would doubt that applies on modern systems.

Reply:
Makes sense, removed. I guess the main possible optimization is persisting updates only before returning.